### PR TITLE
Add UDP prober.

### PR DESCRIPTION
**Pull request by KalivarapuReshma**
For issue #184.

### CODE REVIEW

File: CONFIGURATION.md [hunk 1]
```diff
@@ -117,6 +117,23 @@ tls_config:
...
 
 ```
+### <udp_probe>
+
+```yml
+
+# The preferred IP protocol of the TCP probe (ip4, ip6).
+[ preferred_ip_protocol: <string> | default = "ip6" ]
+
+# The query sent in the TCP probe and the expected associated response.
```
Comment (thz) [inline CONFIGURATION.md]: comments are talking about TCP.

File: CONFIGURATION.md [hunk 2]
```diff
@@ -117,6 +117,23 @@ tls_config:
...
+# starttls upgrades TCP connection to TLS.
+query_response:
+  [ - [ [ expect: <string> ],
+        [ send: <string> ],
+        [ starttls: <boolean | default = false> ]
```
Comment (thz) [inline CONFIGURATION.md]: starttls is not implemented (as advertised)

File: prober/udp.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	registry.MustRegister(probeFailedDueToRegex)
+	deadline, _ := ctx.Deadline()
+
+	conn, err := dialUDP(ctx, target, module, registry, logger)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error dialing UDP", "err", err)
+		return false
+	}
+	defer conn.Close()
+	level.Info(logger).Log("msg", "Successfully dialed")
```
Comment (brian-brazil) [inline prober/udp.go]: Does this mean much with UDP to an IP address? I can see those not familiar with how socket APIs work for UDP getting confused by this log line and thinking it means the packet was sent.

File: prober/udp.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		level.Info(logger).Log("msg", "Processing query response entry", "entry_number", i)
+		send := qr.Send
+		if qr.Expect != "" {
+			re, err := regexp.Compile(qr.Expect)
+			if err != nil {
+				level.Error(logger).Log("msg", "Could not compile into regular expression", "regexp", qr.Expect, "err", err)
+				return false
+			}
+			var match []int
+			// Read lines until one of them matches the configured regexp.
```
Comment (brian-brazil) [inline prober/udp.go]: I'm not sure this makes sense for UDP. How many text-based UDP protocols are there?

File: prober/udp.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+	// Set a deadline to prevent the following code from blocking forever.
+	// If a deadline cannot be set, better fail the probe by returning an error
+	// now rather than blocking forever.
+	if err := conn.SetDeadline(deadline); err != nil {
+		level.Error(logger).Log("msg", "Error setting deadline", "err", err)
+		return false
+	}
+
+	scanner := bufio.NewScanner(conn)
```
Comment (brian-brazil) [inline prober/udp.go]: What does this do for UDP?

File: prober/udp.go (hunk 4 above)
Comment (erichorst) [reply prober/udp.go]: I have one I'm interested in using blackbox_exporter for. It makes as much sense for UDP as it does for TCP I think.

**Comment by erichorst** [comment]
I have an interest in this pull request and will test the udp functionality sometime soon and offer some supporting feedback.

File: prober/udp.go (hunk 4 above)
Comment (brian-brazil) [reply prober/udp.go]: I suspect that's a very small proportion of UDP use cases

**Comment by ac427** [comment]
not related to this PR. I am going to ask in some go stack overflow. I found this issue while working on similar PR. The connection passes for all UDP ports ? 

```
$cat foo.go 
package main

import "net"
import "fmt"

func main() {

	conn, err := net.Dial("udp", "0.north-america.pool.ntp.org:80")
	if err != nil {
		fmt.Printf("Fail")
	}
	if conn != nil {
		fmt.Printf("Pass")
	}
}
$go run foo.go 
Pass$
$sudo nmap -sT -sU 0.north-america.pool.ntp.org
[sudo] password for ac: 

Starting Nmap 6.40 ( http://nmap.org ) at 2019-03-08 15:51 EST
Nmap scan report for 0.north-america.pool.ntp.org (216.229.4.69)
Host is up (0.038s latency).
Other addresses for 0.north-america.pool.ntp.org (not scanned): 45.56.118.161 45.33.48.4 45.79.1.70
rDNS record for 216.229.4.69: triton.ellipse.net
Not shown: 1000 filtered ports, 999 open|filtered ports
PORT    STATE SERVICE
123/udp open  ntp
```

**Comment by brian-brazil** [comment]
You didn't send any packets.

**Comment by ac427** [comment]
There is no open port 80 on 0.north-america.pool.ntp.org to send. Isn't it supposed to timeout with error when failed? or at least hang ( that is what it does for tcp ) 

```
$cat foo.go 
package main

import "net"
import "fmt"

func main() {

	conn, err := net.Dial("tcp", "google.com:123")
	if err != nil {
		fmt.Printf("Fail")
	}
	if conn != nil {
		fmt.Printf("Pass")
	}
}
$go run foo.go 
^Csignal: interrupt
```

**Comment by thz** [comment]
@ac427 that is exactly the difference to UDP. You can send a packet and that is it. There is no connection to be established like in TCP. You could wait for a reply packet of course. But then you should start to define the expectation.

**Comment by ac427** [comment]
@thz  , ACK :) .  I feel dump now. Long day!

**Comment by brian-brazil** [comment]
This has gone stale, and it remains unclear how we'd make this work in principle due to the connectionless and usually binary nature of UDP.

**Pull request closed without merging**